server/container/core/src/main/java/org/apache/james/server/core/BufferedDeferredFileOutputStream.java

```diff
@@ -0,0 +1,278 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.james.server.core;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.output.ByteArrayOutputStream;
+import org.apache.commons.io.output.DeferredFileOutputStream;
+import org.apache.commons.io.output.ThresholdingOutputStream;
+
+import java.io.*;
+
+
+/**
+ * An almost copy of {@link DeferredFileOutputStream} with buffered file stream.
```

Comment:
This is IMO a bit sad...

Couldn't we avoid this copy-paste by wrapping by Bufferred streams where needed?

Reply:
maybe taking a OutputStreamSupplier as a parameter would allow to decouple the concerns ?

Reply:
My observation was that MimeMessageInputStreamSource uses some method from DeferredFileOutputStream (`isInMemory/getData/getFile`) directly and wrapping with buffered streams would use an extra buffer even the input is so small that it fits buffer. So consider this extra buffer and extra logic to flush buffer, I was thinking not worth it.

Now I think it should just only need buffering during MimeMessageInputStreamSource's ctor and flush it after `IOUtils.copy` should do the work.